### DISCUSSION TITLE
Working with Schemas

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#335] [Maintainer discussions]
The specs related to issue #170 are ready for review. Please take a look at the attached document and follow the [review guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines).

## Document Link
https://wiki.mathesar.org/en/design/specs/schemas

@mathemancer @kgodey @pavish, your feedback is required.
The deadline for feedback is set to a week from now, the first review round will be completed by Friday 9th. Please plan your review so that at least one round of changes can be made and reviewed if needed.

**Comment by kgodey** [comment]
This looks good, @ghislaineguerin! Here's my feedback:

1. I like the new top navigation. I think we should have an explicit "Schemas" link in the top navigation though, rather than relying on them clicking on (what I assume is) the Mathesar brand icon.
2. The workflow for deleting schemas should be different. Empty schemas can be deleted without confirmation, but if there's a schema that's referenced by another schema, or contains tables, we should show an error message saying "Deleting the schema  will result in X, Y, Z also being deleted (or) X, Y, Z fields will be removed" and ask them to confirm that they want to make all those changes as well.
3. I think there should be some sort of explanation related to the lock icon.
4. We should show a warning when the user tries to delete the last schema in a database, and let them know they will not be able to use the database for anything.

**Reply by ghislaineguerin** [reply]
Thanks, Kriti. I've pushed a new version of the prototype that:
- Let's users navigate to all schemas by clicking on the database icon rather than Mathesar's brand for additional clarity.
- Addresses how non-empty schema is deleted
- Provides tooltip for the lock icon information

Regarding the deletion of all schemas, does this mean we'll let users delete their default/public database? If so, how would they do that from the UI? or is it something they'd only do externally.

**Comment by pavish** [comment]
@ghislaineguerin I think it looks good. I do have some thoughts:

1. I think we can move the database navigation menu to the top bar. Here are my reasons:
   - Both the menus almost do the same operation with regards to viewing and switching, but the new one offers a full page view with additional controls for adding, editing and deleting schemas. 
   - We can try to combine both, since it seems a bit redundant. 
   - We could probably have a settings icon within the database navigation menu (or next to it), which expands the menu into the full screen view.
   - If we choose to not combine both, we could still move the db navigation menu on top. It makes more sense to keep it on the topbar, rather than the sidebar.
2. I agree with @kgodey's views on deleting schema, and the lock icon. I assume the lock icons means that the schema cannot be edited. We could show that on a tooltip upon hover.

**Comment by mathemancer** [comment]
The only additions I have to what others have mentioned is that the `public` schema is special in PostgreSQL, and this should be shown somehow.  You can't really (without breaking things) alter permissions on the `public` schema.  Also, renaming it would be extremely unwise.  Basically, there's a `PUBLIC` user group in PostgreSQL that all users with access to a DB are members of.  Anyone in this group has access to the `public` schema.  You can't remove users from that group, and you can't revoke access to the `public` schema for that group.

**Comment by ghislaineguerin** [comment]
@mathemancer @pavish @kgodey 
I have updated the document to introduce search-based navigation on the top bar and changes to how schemas are deleted. Please review.

**Reply by kgodey** [reply]
@ghislaineguerin there seems to be an issue with the prototype, it's only one screen and I can't click on anything or see anything related to deleting schemas.

Also, has @mathemancer's feedback about denoting the public schema differently been addressed yet?

**Reply by ghislaineguerin** [reply]
@kgodey I've fixed the prototype now. 
As for @mathemancer 's feedback, I do have a different style for the public schema, but I'm working on additional icon updates to make it more clear.

**Comment by kgodey** [comment]
I think the new prototype looks good, thanks!

My only question is about the "Search or jump to..." behavior. Will we be showing a list of most recently used schemas to jump to? What will the search behavior / list of search results look like?

**Reply by ghislaineguerin** [reply]
@kgodey I will expand on this and provide an example with the search as well.

**Comment by pavish** [comment]
@ghislaineguerin
* I'm not seeing a place where all the databases are listed on the new prototype. The only way where we can change the database is using the "Search of jump to..." option. We would probably need a way to list all databases, similar to how we are listing schemas.
* Are sidebar changes part of the same design spec? If so, we would need an option to show/hide the sidebar.

**Reply by kgodey** [reply]
We do need a way for users to figure out which database they are working in and switch to a different database. This was solved in https://github.com/centerofci/mathesar/issues/172, but it seems like the designs for those are obsolete. 

@ghislaineguerin I assume you don't want to go into too much more research in this spec. For now, maybe we can reintroduce the database switching dropdown and use the "Search or jump to..." only to search within schemas in the given database?

**Reply by kgodey** [reply]
@pavish I talked with @ghislaineguerin on Matrix and I've reopened the multiple database switching design issue #172.  Based on the conversation, I think it makes sense to move the "multiple databases" milestone from milestone 4 to milestone 11 so that we can design how to work within a single database well before spending design time on figuring out how to switch databases since it won't be a very common action.

**Reply by pavish** [reply]
@kgodey We can move multiple database support milestone from 4 to 11. I think we can stick to "Editable Tables" as milestone 4, as it was earlier, and pick up Working with Schemas as milestone 5.

**Reply by kgodey** [reply]
Sounds good.

**Comment by mathemancer** [comment]
I think the modification to the way the `public` schema looks good.

**Comment by ghislaineguerin** [comment]
@kgodey @pavish @mathemancer 
Review notes have been added to the document and listed below for reference.

## Search and Jump To Navigation
With the introduction of the top navigation, we identified the opportunity to have a global control for navigation across databases and schemas that do not rely on the sidebar being visible at all times. This change would simplify other aspects of the user experience, allowing users to navigate away from a schema while keeping track of global notifications and other async operations.
When activated, the search and jump to navigation should display a list of recently navigated objects (schemas and databases).

## Public Schemas
Public schemas within Mathesar will be considered to be a special type. The system will add Contraints to prevent deletion or edits that might disrupt the database functionality.

The public schemas will also have a distinct visual icon to identify them.

## List of Databases
Changes introduced on this spec have invalidated the previously proposed solutions for database navigation. We will revisit the issues on this functionality in a later milestone.

**Comment by mathemancer** [comment]
After chatting with @eito-fis and thinking a bit more deeply about the schema deletion process, I think there are a couple of questions unanswered:
- Do we intend to allow `CASCADE` behavior?  This would let a user delete a schema and anything that depends on it, but it's quite aggressive.  
- If we don't allow `CASCADE`, there isn't a good way (I can think of) to automate dealing with dependencies.  Are we hoping to provide guidance to the user beyond some list of dependent objects?

**Reply by kgodey** [reply]
@mathemancer Yes, we will allow `CASCADE` behavior, that's what deleting a schema means. The idea is that we'll show the user everything that will be deleted and they can choose to delete everything or not do anything. We can figure out more nuanced behavior later (after the MVP).
